subprojects/plugins/src/integTest/groovy/org/gradle/api/tasks/JavaExecDebugIntegrationTest.groovy

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks
+
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.jvm.JDWPUtil
+import org.gradle.internal.os.OperatingSystem
+import org.gradle.test.fixtures.ConcurrentTestUtil
+import org.junit.Rule
+import spock.lang.IgnoreIf
+
+class JavaExecDebugIntegrationTest extends AbstractIntegrationSpec {
+
+    @Rule
+    JDWPUtil debugClient = new JDWPUtil()
+
+    def "Debug is disabled by default"(String taskName) {
+        setup:
+        sampleProject"""
+            debugOptions {
+            }
+        """
+
+        expect:
+        succeeds(taskName)
+
+        where:
+        taskName << ['runJavaExec', 'runProjectJavaExec', 'test']
+    }
+
+    def "Debug mode can be disabled"(String taskName) {
+        setup:
+        sampleProject"""
+            debugOptions {
+                enabled = false
+            }
+        """
+
+        expect:
+        succeeds(taskName)
+
+        where:
+        taskName << ['runJavaExec', 'runProjectJavaExec', 'test']
+    }
+
+    def "Debug session fails without debugger"(String taskName) {
+        setup:
+        sampleProject"""
+            debugOptions {
+                enabled = true
+                server = false
+            }
+        """
+
+        expect:
+        def failure = executer.withTasks(taskName).withStackTraceChecksDisabled().runWithFailure()
+        failure.assertHasErrorOutput('ERROR: transport error 202: connect failed: Connection refused')
+
+        where:
+        taskName << ['runJavaExec', 'runProjectJavaExec', 'test']
+    }
+
+    def "Can debug Java exec with socket listen type debugger (server = false)"(String taskName) {
```

Comment:
❓ Should we use `@Unroll` annotation here? I never really know when we shouldn't use that annotation. :-(

Reply:
It's a bit of a taste thing. With the annotation present, you can't rerun specific tests from the IDE's test result view (as the method name is adjusted).